urwid/_async_kw_event_loop.py

```diff
@@ -0,0 +1,244 @@
+                the loop is started and force it to create a nursery on its
+                own.
+        """
+        import trio
+
+        self._idle_handle = 0
+        self._idle_callbacks = {}
+        self._pending_tasks = []
+
+        self._trio = trio
+        self._nursery = None
+
+        self._sleep = trio.sleep
+        self._wait_readable = trio.hazmat.wait_readable
+
+    def alarm(self, seconds, callback):
+        """Calls `callback()` a given time from now.  No parameters are passed
+        to the callback.
+
+        Parameters:
+            seconds: time in seconds to wait before calling the callback
+            callback: function to call from the event loop
+
+        Returns:
+            a handle that may be passed to `remove_alarm()`
+        """
+        return self._start_task(self._alarm_task, seconds, callback)
+
+    def enter_idle(self, callback):
+        """Calls `callback()` when the event loop enters the idle state.
+
+        There is no such thing as being idle in a Trio event loop so we
+        simulate it by repeatedly calling `callback()` with a short delay.
+        """
+        self._idle_handle += 1
+        self._idle_callbacks[self._idle_handle] = callback
+        return self._idle_handle
+
+    def remove_alarm(self, handle):
+        """Removes an alarm.
+
+        Parameters:
+            handle: the handle of the alarm to remove
+        """
+        return self._cancel_scope(handle)
+
+    def remove_enter_idle(self, handle):
+        """Removes an idle callback.
+
+        Parameters:
+            handle: the handle of the idle callback to remove
+        """
+        try:
+            del self._idle_callbacks[handle]
+        except KeyError:
+            return False
+        return True
+
+    def remove_watch_file(self, handle):
+        """Removes a file descriptor being watched for input.
+
+        Parameters:
+            handle: the handle of the file descriptor callback to remove
+
+        Returns:
+            True if the file descriptor was watched, False otherwise
+        """
+        return self._cancel_scope(handle)
+
+    def _cancel_scope(self, scope):
+        """Cancels the given Trio cancellation scope.
+
+        Returns:
+            True if the scope was cancelled, False if it was cancelled already
+            before invoking this function
+        """
+        scope, cancelled = scope
+        existed = not cancelled[0]
+        cancelled[0] = True
+        scope.cancel()
+        return existed
+
+    def _create_cancel_scope(self):
+        """Creates a Trio cancellation scope and a corresponding mutable flag
+        that indicates whether the scope was cancelled already _from_ urwid.
+
+        This is needed because `CancelScope.cancelled_caught` stays `False` until
+        someone actually _handles_ the cancellation, and
+        `CancelScope.cancel_called` can only be called from an async context
+        (which we cannot guarantee).
```

Comment:
Huh, does urwid use threads? "async context" just means "inside the same thread where `trio.run` is running".

Reply:
> It could be the case that the unit test is doing something fishy (e.g., not actually running the loop but just adding and removing a handle outside the loop), but since this part of the unit test is common for all event loop tests within urwid, it seemed easier not to argue about changing the unit test itself.

I see, yeah, that must be what's going on. If the loop was running then `cancel_called` would work.

It sounds like there's an argument for fixing the test, rather than adding contortions to the main loop to work around these issues, but I can see why you might not want to argue about it too...

Reply:
> It sounds like there's an argument for fixing the test, rather than adding contortions to the main loop to work around these issues

@tonycpsu @ulidtko Not sure who calls the shots in issues like this; would you be open to changing the unit tests such that the tests for the `remove_alarm()` and `remove_watch_file()` calls (in `test_remove_alarm()` and `test_remove_watch_file()`, respectively) are also executed _while_ the event loop is running? The reason is that right now in the implementation of the Trio event loop we need to jump through extra hoops to make these methods work even _outside_ the event loop. (For me it's okay to keep it this way - after all, it works - I just need a second opinion).

Reply:
@ntamas : I agree with @njsmith that fixing the test code is preferable to complicating the run loop code.  As long as the test code changes don't cause any test failures when testing with the other run loops, I don't have a problem with it.  If it does, then those would need to be addressed, and at that point maybe we just accept what you have that's working now and live to fight another day.

Reply:
Thanks for the feedback - I'll give it a go to refactor the tests and simplify the run loop and I'll push the changes here if it works smoothly.

Reply:
PR updated - I have changed the tests a bit and it now works without additional hackery in the TrioEventLoop code.

Note that I needed to changed the tests for the TwistedEventLoop as well because its reactor cannot be restarted.